src/Analyser/MutatingScope.php

```diff
@@ -3436,6 +3474,26 @@ public function specifyExpressionType(Expr $expr, Type $type, Type $nativeType):
 		$nativeTypes = $scope->nativeExpressionTypes;
 		$nativeTypes[$exprString] = ExpressionTypeHolder::createYes($expr, $nativeType);
 
+		foreach ($scope->conditionalExpressions as $conditionalExprString => $conditionalExpressions) {
+			foreach (array_reverse($conditionalExpressions) as $conditionalExpression) {
+				$newConditionExpressionTypeHolders = [];
+				foreach ($conditionalExpression->getConditionExpressionTypeHolders() as $holderExprString => $conditionalTypeHolder) {
+					if ($holderExprString === $exprString && $expressionTypes[$holderExprString]->equals($conditionalTypeHolder)) {
+						continue;
+					}
+					$newConditionExpressionTypeHolders[$holderExprString] = $conditionalTypeHolder;
+				}
+				if ($newConditionExpressionTypeHolders === []) {
+					if ($conditionalExpression->getTypeHolder()->getCertainty()->no()) {
+						unset($expressionTypes[$conditionalExprString]);
+					} else {
+						$expressionTypes[$conditionalExprString] = $conditionalExpression->getTypeHolder();
+					}
+					continue 2;
+				}
+			}
+		}
```

Comment:
I'm not satisfied enough with this implementation, but this works at least.
I think after conditional types merging logic has improved, we can delete matching ConditionExpressionTypeHolders